File: java/src/main/java/sk/epostak/sdk/resources/FirmsResource.java
package sk.epostak.sdk.resources;

import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.*;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Manage client firms (integrator keys).
 * <p>
 * Provides operations for listing, inspecting, and assigning firms,
 * as well as registering Peppol identifiers. Requires an integrator key
 * ({@code sk_int_*}) for most operations.
 * <p>
 * Access via {@code client.firms()}.
 *
 * <pre>{@code
 * // List all firms, then get detail for each
 * List<FirmSummary> firms = client.firms().list();
 * for (FirmSummary firm : firms) {
 *     FirmDetail detail = client.firms().get(firm.id());
 * }
 * }</pre>
 */
public final class FirmsResource {

    private final HttpClient http;

    /**
     * Creates a new firms resource.
     *
     * @param http the HTTP client used for API communication
     */
    public FirmsResource(HttpClient http) {
        this.http = http;
    }

    /**
     * List all firms accessible to the current API key.
     *
     * @return list of firm summaries, or an empty list if none
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public List<FirmSummary> list() {
        FirmsListWrapper wrapper = http.get("/firms", FirmsListWrapper.class);
        return wrapper != null ? wrapper.firms() : List.of();
    }

    /**
     * Get detailed information for a specific firm, including Peppol identifiers.
     *
     * @param id the firm UUID
     * @return the firm detail
     * @throws sk.epostak.sdk.EPostakException if the firm is not found or the request fails
     */
    public FirmDetail get(String id) {
        return http.get("/firms/" + HttpClient.encode(id), FirmDetail.class);
    }

    /**
     * List documents for a specific firm with filtering and pagination.
     *
     * @param id        the firm UUID
     * @param offset    pagination offset (0-based), or {@code null} for default
     * @param limit     max results per page, or {@code null} for default
     * @param direction filter by direction: {@code "inbound"} or {@code "outbound"}, or {@code null} for all
     * @return paginated list of documents belonging to the firm
     * @throws sk.epostak.sdk.EPostakException if the firm is not found or the request fails
     */
    public InboxListResponse documents(String id, Integer offset, Integer limit, String direction) {
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("offset", offset);
        params.put("limit", limit);
        params.put("direction", direction);
        return http.get("/firms/" + HttpClient.encode(id) + "/documents" + HttpClient.buildQuery(params), InboxListResponse.class);
    }

    /**
     * List documents for a specific firm with default parameters.
     *
     * @param id the firm UUID
     * @return paginated list of documents belonging to the firm
     * @throws sk.epostak.sdk.EPostakException if the firm is not found or the request fails
     */
    public InboxListResponse documents(String id) {
        return documents(id, null, null, null);
    }

    /**
     * Register a Peppol identifier for a firm, enabling it to send and receive
     * documents on the Peppol network.
     *
     * <pre>{@code
     * PeppolIdentifierResponse resp = client.firms().registerPeppolId(
     *     "firm_uuid", "0245", "12345678");
     * }</pre>
     *
     * @param id         the firm UUID
     * @param scheme     identifier scheme, e.g. {@code "0245"} for Slovak DIČ
     * @param identifier identifier value, e.g. {@code "12345678"}
     * @return the registration response with the full Peppol ID and timestamp
     * @throws sk.epostak.sdk.EPostakException if the firm is not found or registration fails
     */
    public PeppolIdentifierResponse registerPeppolId(String id, String scheme, String identifier) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("scheme", scheme);
        body.put("identifier", identifier);
        return http.post("/firms/" + HttpClient.encode(id) + "/peppol-identifiers", body, PeppolIdentifierResponse.class);
    }

    /**
     * Assign a firm to this integrator by ICO (company registration number).
     *
     * @param ico the Slovak company registration number (ICO)
     * @return the assignment response with firm details and status
     * @throws sk.epostak.sdk.EPostakException if the ICO is invalid or the request fails
     */
    public AssignFirmResponse assign(String ico) {
        Map<String, Object> body = Map.of("ico", ico);
        return http.post("/firms/assign", body, AssignFirmResponse.class);
    }

    /**
     * Batch assign firms to this integrator by ICO (max 50 per request).
     *
     * @param icos list of Slovak company registration numbers (ICO), max 50
     * @return the batch response with per-ICO results and error details
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public BatchAssignResponse assignBatch(List<String> icos) {
        Map<String, Object> body = Map.of("icos", icos);
        return http.post("/firms/assign/batch", body, BatchAssignResponse.class);
    }

    // -- internal wrapper for the list response --------------------------------

    private record FirmsListWrapper(List<FirmSummary> firms) {}

    /**
     * Response from registering a Peppol ID for a firm. Returned as HTTP 201.
     *
     * @param peppolId           the full Peppol participant ID, e.g. {@code "0245:1234567890"}
     * @param registrationStatus registration status, typically {@code "pending"} — SMP
     *                           publication completes asynchronously within 24 hours
     * @param message            human-readable confirmation message
     */
    public record PeppolIdentifierResponse(
            String peppolId,
            String registrationStatus,
            String message
    ) {}
}

File: dotnet/src/EPostak/Resources/FirmsResource.cs
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>
/// Manage client firms and their Peppol identifiers.
/// For direct API keys (<c>sk_live_*</c>), this returns your own firm.
/// For integrator keys (<c>sk_int_*</c>), this manages all assigned client firms.
/// </summary>
public sealed class FirmsResource
{
    private readonly HttpRequestor _http;

    internal FirmsResource(HttpRequestor http) => _http = http;

    /// <summary>
    /// List all firms accessible by the current API key.
    /// For integrator keys, returns all assigned client firms.
    /// For direct keys, returns just the firm linked to the key.
    /// </summary>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>List of firm summaries with Peppol registration status.</returns>
    /// <example>
    /// <code>
    /// var firms = await client.Firms.ListAsync();
    /// foreach (var firm in firms)
    ///     Console.WriteLine($"{firm.Name} (ICO: {firm.Ico}) - Peppol: {firm.PeppolStatus}");
    /// </code>
    /// </example>
    public async Task<List<FirmSummary>> ListAsync(CancellationToken ct = default)
    {
        var res = await _http.RequestAsync<FirmsListResponse>(HttpMethod.Get, "/firms", ct).ConfigureAwait(false);
        return res.Firms;
    }

    /// <summary>
    /// Get detailed information about a firm including address, tax identifiers,
    /// and all registered Peppol identifiers.
    /// </summary>
    /// <param name="id">The firm UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Full firm details including Peppol identifiers and registration date.</returns>
    /// <example>
    /// <code>
    /// var firm = await client.Firms.GetAsync("firm_abc123");
    /// Console.WriteLine($"{firm.Name}, ICO: {firm.Ico}, Peppol: {firm.PeppolId}");
    /// foreach (var pid in firm.PeppolIdentifiers)
    ///     Console.WriteLine($"  {pid.Scheme}:{pid.Identifier}");
    /// </code>
    /// </example>
    public Task<FirmDetail> GetAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<FirmDetail>(HttpMethod.Get, $"/firms/{Uri.EscapeDataString(id)}", ct);

    /// <summary>
    /// List documents (inbound and/or outbound) for a specific firm.
    /// Useful for integrators to browse a client firm's document history.
    /// </summary>
    /// <param name="id">The firm UUID.</param>
    /// <param name="params">Optional filters: pagination and direction (inbound/outbound).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Paginated list of documents for the specified firm.</returns>
    /// <example>
    /// <code>
    /// var docs = await client.Firms.DocumentsAsync("firm_abc123", new FirmDocumentsParams
    /// {
    ///     Direction = DocumentDirection.Inbound,
    ///     Limit = 20
    /// });
    /// Console.WriteLine($"Found {docs.Total} inbound documents");
    /// </code>
    /// </example>
    public Task<InboxListResponse> DocumentsAsync(string id, FirmDocumentsParams? @params = null, CancellationToken ct = default)
    {
        var direction = @params?.Direction switch
        {
            DocumentDirection.Inbound => "inbound",
            DocumentDirection.Outbound => "outbound",
            _ => null
        };
        var qs = HttpRequestor.BuildQuery(
            ("offset", @params?.Offset?.ToString()),
            ("limit", @params?.Limit?.ToString()),
            ("direction", direction));
        return _http.RequestAsync<InboxListResponse>(HttpMethod.Get, $"/firms/{Uri.EscapeDataString(id)}/documents{qs}", ct);
    }

    /// <summary>
    /// Register a Peppol identifier for a firm on the SMP (Service Metadata Publisher).
    /// This makes the firm discoverable and able to receive documents on the Peppol network.
    /// Common schemes: <c>0192</c> (Slovak ICO), <c>0088</c> (EAN/GLN).
    /// </summary>
    /// <param name="id">The firm UUID.</param>
    /// <param name="scheme">The Peppol identifier scheme (e.g. "0192" for Slovak ICO).</param>
    /// <param name="identifier">The identifier value within the scheme (e.g. the ICO number "12345678").</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The registered Peppol ID, scheme, identifier, and registration timestamp.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Firms.RegisterPeppolIdAsync("firm_abc123", "0192", "12345678");
    /// Console.WriteLine($"Registered: {result.PeppolId} at {result.RegisteredAt}");
    /// </code>
    /// </example>
    public Task<PeppolIdentifierResponse> RegisterPeppolIdAsync(string id, string scheme, string identifier, CancellationToken ct = default)
        => _http.RequestAsync<PeppolIdentifierResponse>(HttpMethod.Post,
            $"/firms/{Uri.EscapeDataString(id)}/peppol-identifiers",
            new { scheme, identifier }, ct);

    /// <summary>
    /// Link this integrator to a Firm that has already completed FS SR signup
    /// and granted consent. <strong>Lookup-only</strong> — this endpoint cannot
    /// create new Firms. The target Firm must have completed FS SR PFS signup
    /// and granted consent to this integrator before the link succeeds.
    /// Integrator keys only.
    /// </summary>
    /// <remarks>
    /// On error, inspect <see cref="EPostakException.Code"/>:
    /// <list type="bullet">
    ///   <item><description><c>FIRM_NOT_REGISTERED</c> (HTTP 404) — no Firm with that ICO exists yet. Direct the firm to complete FS SR PFS signup before retrying.</description></item>
    ///   <item><description><c>CONSENT_REQUIRED</c> (HTTP 403) — Firm exists but has not granted consent for this integrator to act on its behalf.</description></item>
    ///   <item><description><c>ALREADY_LINKED</c> (HTTP 409) — the integrator already has an active link to this Firm.</description></item>
    /// </list>
    /// </remarks>
    /// <param name="ico">Slovak business registration number (ICO), e.g. "12345678".</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The linked firm details and link status.</returns>
    /// <exception cref="EPostakException">When the firm is not registered, consent is missing, the link already exists, or the request fails.</exception>
    /// <example>
    /// <code>
    /// try
    /// {
    ///     var result = await client.Firms.AssignAsync("12345678");
    ///     Console.WriteLine($"Firm: {result.Firm.Name} ({result.Status})");
    /// }
    /// catch (EPostakException ex) when (ex.Code == "FIRM_NOT_REGISTERED")
    /// {
    ///     // ask the firm to complete FS SR PFS signup first
    /// }
    /// </code>
    /// </example>
    public Task<AssignFirmResponse> AssignAsync(string ico, CancellationToken ct = default)
        => _http.RequestAsync<AssignFirmResponse>(HttpMethod.Post, "/firms/assign", new { ico }, ct);

    /// <summary>
    /// Assign multiple firms by ICO in a single batch request (max 50).
    /// Each ICO is processed independently -- partial failures don't block other assignments.
    /// Integrator keys only.
    /// </summary>
    /// <param name="icos">Collection of Slovak ICO numbers to assign (max 50).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Individual results for each ICO including firm details or error messages.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Firms.AssignBatchAsync(new[] { "12345678", "87654321", "11111111" });
    /// foreach (var r in result.Results)
    /// {
    ///     if (r.Error is not null)
    ///         Console.WriteLine($"ICO {r.Ico}: FAILED - {r.Error}");
    ///     else
    ///         Console.WriteLine($"ICO {r.Ico}: {r.Firm?.Name} ({r.Status})");
    /// }
    /// </code>
    /// </example>
    public Task<BatchAssignFirmsResponse> AssignBatchAsync(IEnumerable<string> icos, CancellationToken ct = default)
        => _http.RequestAsync<BatchAssignFirmsResponse>(HttpMethod.Post, "/firms/assign/batch", new { icos }, ct);
}
